### PR TITLE
init programs.gns3

### PR DESCRIPTION
**Pull request by ghost**
## Description of changes

GNS3 is already a package, but it depends on a few other packages in order to function properly. These are now managed by the programs.gns3 option.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/programs/gns3.nix [hunk 1]
```diff
@@ -0,0 +1,110 @@
+{ pkgs, config, lib, ... }:
+
+with lib;
```
Comment (AndersonTorres) [inline nixos/modules/programs/gns3.nix:3]:
```suggestion
```

File: nixos/modules/programs/gns3.nix [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+      cfg.serverPackage
+      cfg.rdcPackage
+      cfg.terminalPackage
+    ];
+
+  };
+
+  meta.maintainers = with maintainers; [ loveisgrief ];
+}
+
```
Comment (AndersonTorres) [inline nixos/modules/programs/gns3.nix:110]:
```suggestion
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: nixos/modules/programs/gns3.nix (hunk 1 above)
Comment (LoveIsGrief) [reply nixos/modules/programs/gns3.nix:3]: That's necessary for the `mkOption`, `types.XXX`, `mkDoc`, etc.

Comment (AndersonTorres) [reply nixos/modules/programs/gns3.nix:3]: No, it is not.
Prefix them or write an explicit `inherit (lib)` listing.

**Comment by h7x4** [comment]
Just wondering, is this a duplicate of #252790?

**Comment by LoveIsGrief** [comment]
What's the difference with "inherit(lib)"? Isn't the effect the same?
Does that mean that " with maintainers; " should also be an "inherit(maintainers)"?BTW, this is taken from the firefox programs module.

**Comment by AndersonTorres** [comment]
> What's the difference with "inherit(lib)"? Isn't the effect the same?

"The effect" is not the only thing that matters - even because `inherit (lib)` is not equivalent to `with lib`.

It is about control and predictability.

If you have some knowledge about Python, `with lib` is similar to `from lib import *`.
You are bringing all the functions from `lib` into the scope of  the whole file.
On the other hand, `inherit (lib) mkOption mdDoc` is similar to `from lib import mkOptions mdDoc`, bringing only the specified functions into the scope.

> Does that mean that " with maintainers; " should also be an "inherit(maintainers)"?

Ideally, `with` should be abolished. Copying the [best practices](https://nix.dev/guides/best-practices.html) guide:

> 1. Static analysis can’t reason about the code, because it would have to actually evaluate the with'ed file to see which names are in scope.
> 2. The scoping rules are not intuitive
> 3. When more than one with is used, it’s not clear anymore where the names are coming from.

> BTW, this is taken from the firefox programs module.

Many failures do not make a success.

https://github.com/NixOS/nixpkgs/issues/208242

**Comment by LoveIsGrief** [comment]
Thanks @AndersonTorres for the explanation. I was not aware of the difference. `nix` is in dire need of a linter like `clippy` for rust that is run on all PRs by the CICD, but that's a topic for another time.

@h7x4 the other PR adds a systemd service for the gns3 server, which is good if you have a separate machine for the gns3 server and want to access it from clients on separate machines running the gns3 GUI client. For a local machine, a constantly running gns3 server is unnecessary.

The usecase for this PR is a user looking to have all dependencies for the GUI client and a local gns3 server.
gns3-gui starts the server when needed and stops it (configured to do so by default) when exited. The server is run with user running gns3-gui.  
The GUI also has additional dependencies (a terminal emulator and telnet, and a remote desktop client for nodes that are accessible with VNC).

File: nixos/modules/programs/gns3.nix (hunk 1 above)
Comment (LoveIsGrief) [reply nixos/modules/programs/gns3.nix]: Done in the last [rebase](https://github.com/NixOS/nixpkgs/compare/ea492c6e07d5d0444982949791ea4f58cf65f071..27f62184229e9e6c505d49e5ee95ef562477598a)

**Pull request closed without merging**